index.bs

```diff
@@ -78,4 +278,322 @@ interface StorageBucket {
+
+    1. [=/Resolve=] |p| with |dictionary|.
+
+1. Return |p|.
+
+</div>
+
+<aside class="note">
+
+The [=StorageBucket/quota value=] will be ignored if it exceeds the total amount of space available to the [=/storage shelf=], i.e. site.
+Its intended use is to keep a specific bucket from using up the entire site's storage space.
+
+</aside>
+
+<h3 id="storage-bucket-durability">Durability</h3>
+
+A [=/storage bucket=] has a <dfn for=StorageBucket>durability value</dfn>, a {{StorageBucketDurability}}.
+The user agent MAY initialize this value to "{{StorageBucketDurability/strict}}" or
+"{{StorageBucketDurability/relaxed}}". The user agent SHOULD ignore the [=StorageBucket/durability value=]
+for operations that otherwise specify durability behavior.
+
+<div algorithm>
+
+The <dfn method for="StorageBucket">durability()</dfn> method steps are:
+
+1. Let |p| be [=a new promise=].
+
+1. Let |bucket| be [=this=]'s [=/storage bucket=].
+
+1. If |bucket|'s [=storage bucket/removed=] flag is true, [=reject=] |p| with an {{InvalidStateError}}.
+
+1. Otherwise, [=/resolve=] |p| with |bucket|'s [=StorageBucket/durability value=].
+
+1. Return |p|.
+
+</div>
+
+<aside class="note">
+
+The durability reflects whether the user agent will prioritize performance or durability when completing operations on data in the bucket.
+Some operations, such as {{IDBTransaction}}s, can specify a durability which will override the [=StorageBucket/durability value=]. The default
+[=StorageBucket/durability value=] is left to the user agent. Durability can be specified when a bucket is initially created, but not changed later.
+The {{StorageBucket/durability()}} method should report what the user agent will actually do. The value is one of the following:
+
+: "{{StorageBucketDurability/strict}}"
+:: The user agent will consider that operations on data in the bucket are successful only after verifying that outstanding changes have been written to a persistent storage medium.
+: "{{StorageBucketDurability/relaxed}}"
+:: The user agent will consider that operations on data in the bucket are successful as soon as all outstanding changes have been written to the operating system, without subsequent verification.
+
+</aside>
+
+<h3 id="storage-bucket-expiration">Expiration</h3>
+
+A [=/storage bucket=] has an <dfn for="StorageBucket">expiration value</dfn>, which is either null or a [=duration=], initially null.
+Specifies the upper limit of a bucket lifetime.
+
+User agents MUST remove any bucket that has expired when {{StorageBucketManager/keys()}} is called, regardless of the [=/bucket mode=].
+User agents MUST remove a bucket that has expired when it is opened via {{StorageBucketManager/open()}}, regardless of the [=/bucket mode=].
+User agents MAY clear buckets whose [=/bucket mode=] is `"best-effort"` before the
+specified timestamp when faced with storage pressure.
+
+<div algorithm>
+
+The <dfn method for="StorageBucket">setExpires(|expires|)</dfn> method steps are:
```

Comment:
I was a bit confused between durations and moments before.  I thought this was taking an absolute time moment, but I see in the text above there was some attempt to use durations.

Should we have `setExpires()` using absolute time moment and then also `setExpiresByDuration()` using a duration value to compute the expires time?

Reply:
Sorry, I think I added to the confusion because I was also confused when writing this draft 😅 .  I think I'm leaning toward updating everything to moments, and having `setExpires()` also use absolute time moment. WDYT?

Reply:
I think that makes sense.  We just have to do some conversion from the value the js passes in for "expires" into moments.  I think if we treat that value as a duration that is added to the unix epoch then we will have a moment we can use for comparisons.

cc @noamr as the expert who I believe wrote the time spec infrastructure.

Reply:
Sorry for my confusion.  @jyasskin, can you help here?

I thought we needed a conversion because `[=current wall time=]` returns the result from `[=coarsen time=]` which is a moment.  Is there something that defines how timestamps compare against moments?  AFAIK this spec PR does not say anywhere that the timestamp is relative to the unix epoch, so it seems ambiguous to me.

Reply:
I would suggest using [=epoch-relative timestamp=] instead of [=current wall time=]

Reply:
Since epoch-relative timestamps are just untyped numbers that get confused between their interpretations as epoch-relative, page-load-relative, and durations between other points in time, if a spec is trying to represent a point in time, I prefer to use moments inside of specifications. Good wording for that, given an `expires` initialized by `Date.now() + duration_ms`, would probably be `Let |expiration| be |expires| milliseconds after the [=Unix epoch=].`

I have less of a strong preference for how to express the Javascript interface, at least until Temporal is ready. I'd lean toward a `setMaxAge(duration_ms)` so readers don't have to wonder if it's a duration or a timestamp, and then `Let |expiration| be |duration_ms| milliseconds after [=this=]'s [=relevant settings object=]'s [=environment settings object/current wall time=].`